client/write_test.go

```diff
@@ -21,30 +21,141 @@
 package client
 
 import (
+	"errors"
 	"sync"
 	"testing"
 
 	"github.com/m3db/m3cluster/shard"
 	"github.com/m3db/m3db/topology"
 
 	"github.com/golang/mock/gomock"
+	tterrors "github.com/m3db/m3db/network/server/tchannelthrift/errors"
+	xerrors "github.com/m3db/m3x/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
+// shard state tests
+
+func testWriteSuccess(t *testing.T, state shard.State, success bool) {
+	var writeWg sync.WaitGroup
+
+	wState, s, host := testSetup(t, &writeWg)
+	setShardStates(t, s, host, state)
+	wState.completionFn(host, nil)
+
+	if success {
+		assert.Equal(t, int32(1), wState.success)
+	} else {
+		assert.Equal(t, int32(0), wState.success)
+	}
+
+	// cleanup
+	wState.decRef() // end introspection
+	writeWg.Wait()  // wait for write to complete
+}
+
 func TestWriteToAvailableShards(t *testing.T) {
-	assert.Equal(t, int32(sessionTestReplicas), shardStateWriteTest(t, shard.Available))
+	testWriteSuccess(t, shard.Available, true)
 }
 
 func TestWriteToInitializingShards(t *testing.T) {
-	assert.Equal(t, int32(0), shardStateWriteTest(t, shard.Initializing))
+	testWriteSuccess(t, shard.Initializing, false)
 }
 
 func TestWriteToLeavingShards(t *testing.T) {
-	assert.Equal(t, int32(0), shardStateWriteTest(t, shard.Leaving))
+	testWriteSuccess(t, shard.Leaving, false)
+}
+
+// retryability test
+
+type errTestFn func(error) bool
+
+func retryabilityCheck(t *testing.T, wState *writeState, writeWg *sync.WaitGroup, testFn errTestFn) {
+	require.True(t, len(wState.errors) == 1)
+	assert.True(t, testFn(wState.errors[0]))
+
+	// cleanup
+	wState.decRef() // end introspection
+	writeWg.Wait()  // wait for write to complete
```

Comment:
Hm feels strange to have common cleanup code here when `testSetup` is broken out and independent.  Also since the functions declared in this file is global to the package `*_test.go` files it might be better to call it `writeTestSetup` and `writeTestTeardown` so its clear where to use them.